pkg/buildkit/worker/leasemanager.go

```diff
@@ -0,0 +1,345 @@
+		return nil, err
+	}
+
+	m.log.Info("No unleased pods found, scaling workload")
+	if pod, err = m.getUnleasedPod(ctx); err != nil {
+		return nil, fmt.Errorf("unable to acquire lease after scaling: %w", err)
+	}
+
+	return pod, nil
+}
+
+func (m *podLeaseManager) getUnleasedPod(ctx context.Context) (*corev1.Pod, error) {
+	m.log.Info("Querying for pods", "namespace", m.namespace, "opts", m.podListOptions)
+	list, err := m.podClient.List(ctx, m.podListOptions)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, pod := range list.Items {
+		if _, ok := pod.Annotations[leasedAnnotation]; !ok {
+			m.log.Info("Found unleased pod", "name", pod.Name, "namespace", pod.Namespace)
+			return &pod, nil
+		}
+	}
+
+	m.log.Info("No unleased pods found")
+	return nil, errNoUnleasedPods
+}
+
+func (m *podLeaseManager) buildURL(ctx context.Context, podName string) (string, error) {
+	m.log.Info("Querying for endpoints", "name", m.serviceName, "namespace", m.namespace)
+	endpoints, err := m.endpointsClient.Get(ctx, m.serviceName, metav1.GetOptions{})
+	if err != nil {
+		return "", err
+	}
+
+	var host string
+
+addressScan:
+	for _, subset := range endpoints.Subsets {
+		for _, address := range subset.Addresses {
+			if address.TargetRef.Name == podName {
+				for _, port := range subset.Ports {
+					if port.Port == m.servicePort {
+						host = strings.Join([]string{address.Hostname, endpoints.Name, endpoints.Namespace}, ".")
+						m.log.Info("Found eligible endpoint address", "host", host)
+
+						break addressScan
+					}
+				}
+			}
+		}
+	}
+
+	if host == "" {
+		return "", fmt.Errorf("endpoints %q does not expose pod %s on port %d", endpoints.Name, podName, 1234)
+	}
+
+	u, err := url.ParseRequestURI(fmt.Sprintf("tcp://%s:%d", host, m.servicePort))
+	if err != nil {
+		return "", fmt.Errorf("failed to parse endpoint url: %w", err)
+	}
+
+	return u.String(), nil
+}
+
+func (m *podLeaseManager) monitorPods(ctx context.Context) {
+	m.log.Info("Starting worker pod monitor", "syncTime", m.podSyncTime.String())
+	ticker := time.NewTicker(m.podSyncTime)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			m.log.Info("Shutting down worker pod monitor")
+			return
+		case <-ticker.C:
+			m.log.Info("Checking worker pool for expired leases")
+			if err := m.terminateUnleasedPods(ctx); err != nil {
+				m.log.Error(err, "Failed to clean worker pool")
+			}
+			m.log.Info("Worker pool lease check complete")
+		}
+	}
+}
+
+func (m *podLeaseManager) terminateUnleasedPods(ctx context.Context) error {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	m.log.Info("Querying for pods", "namespace", m.namespace, "opts", m.podListOptions)
+	list, err := m.podClient.List(ctx, m.podListOptions)
+	if err != nil {
+		return err
+	}
+
+	var removals []string
+	for idx := len(list.Items) - 1; idx >= 0; idx-- {
+		pod := list.Items[idx]
+		log := m.log.WithValues("podName", pod.Name)
+
+		if id, ok := pod.Annotations[managerIDAnnotation]; ok && id != m.uuid {
```

Comment:
When would the manager ID annotation exist and be different than expected? Or just defensive coding?

Reply:
It was a little difficult to think about leasing/unleasing in the scope of controller restarts so this is how I accomplished it.

### Metadata

Manager ID is added in `Lease` and removed in `Release.

https://github.com/dominodatalab/hephaestus/blob/7f6c17bdcd6db509f8ded5e5550c88729de22d36/pkg/buildkit/worker/leasemanager.go#L120-L123

https://github.com/dominodatalab/hephaestus/blob/7f6c17bdcd6db509f8ded5e5550c88729de22d36/pkg/buildkit/worker/leasemanager.go#L160-L164

### Predicates

If the controller restarts for some reason (terminated manually, by k8s, or upgrade), then we need to ensure the pods are cleaned up:

1. Pods that were released will NOT have a manager ID and the worker pool GC routine will honor their expiry time.
2. Pods that were leased will have both a lease and manager ID annotation. Since the new lease manager instance will have a new UUID, it will mark these for termination.
3. Pods with invalid metadata (no manager ID, no expiry time) will be marked for termination.

Reply:
Does this make sense now? Would you change anything?

Reply:
I didn't think of this case:
> Pods that were leased will have both a lease and manager ID annotation. Since the new lease manager instance will have a new UUID, it will mark these for termination.

That makes sense, we can't track builds after a restart.